datalad_next/processors/decode_processor.py

```diff
@@ -0,0 +1,75 @@
+""" Data processor that decodes bytes into strings """
+
+from __future__ import annotations
+
+from typing import (
+    Generator,
+    Iterable,
+)
+
+
+__all__ = ['decode_processor']
+
+
+def decode_processor(iterable: Iterable[bytes],
+                     encoding: str = 'utf-8',
+                     backslash_replace: bool = True
+                     ) -> Generator[str, None, None]:
+    """ Decode bytes yielded by ``iterable`` into strings
+
+    Parameters
+    ----------
+    iterable: Iterable[bytes]
+        The iterable that yields bytes that should be decoded
+    encoding: str (default: ``'utf-8'``)
+        The encoding that is used to decode the bytes yielded by ``iterable``
+    backslash_replace: bool (default: ``True``)
+        If ``True``, backslash-escapes are used for undecodable bytes. If
+        ``False``, a ``UnicodeDecodeError`` is raised if a byte sequence cannot
+        be decoded.
+
+    Yields
+    ------
+    str
+        The decoded strings that are generated by decoding the data yielded by
+        ``iterable`` with the specified ``encoding``
+
+    Raises
+    ------
+    UnicodeDecodeError
+        If ``backslash_replace`` is ``False`` and the data yielded by
+        ``iterable`` cannot be decoded with the specified ``encoding``
+    """
+    joined_data = b''
+    position = 0
+    for chunk in iterable:
+        joined_data += chunk
+        while position < len(joined_data):
+            try:
+                yield joined_data[position:].decode(encoding)
+                joined_data = b''
+            except UnicodeDecodeError as e:
+                # If an encoding error occurs, we first check whether it was
+                # in the middle of `joined_data` or whether it extends until the
+                # end of `joined_data`.
+                # If it occurred in the middle of
+                # `joined_data`, we replace it with backslash encoding or
+                # re-raise the decoding error.
+                # If it occurred at the end of `joined_data`, we wait for the
+                # next chunk, which might fix the problem.
```

Comment:
This cannot be a comment. It is essential for understanding what this tool does. It should be moved to the documentation.

In general, I wonder what the use case is for this helper? In yields decodeable chunks that are similiar, but not identical to the items of the input iterator.

Users may have similar concerns, and adding a clear recommendation when this would be useful should be in the documentation. I would think that it would be simpler to re-chunk a bytes "stream", and decode actual items -- rather than trial-and-error decoding.

Reply:
> This cannot be a comment. It is essential for understanding what this tool does. It should be moved to the documentation.

It is documented in the interface documentation for the `backslash_replace`-argument:

```
    backslash_replace: bool (default: ``True``)
        If ``True``, backslash-escapes are used for undecodable bytes. If
        ``False``, a ``UnicodeDecodeError`` is raised if a byte sequence cannot
        be decoded.
```

> In general, I wonder what the use case is for this helper? In yields decodeable chunks that are similiar, but not identical to the items of the input iterator.

The use case is to properly decode bytes into unicode characters, even if the encoding crosses a chunk-border. The simple `map(decode, chunk_source)` would not work, if the encoded unicode character is split between two chunks.
The function `decode_processor` fulfills an equivalent function as `datalad.runner.utils.AssemblingDecoderMixIn`. Tha latter was introduced to fix datalad's issue https://github.com/datalad/datalad/issues/6526.

In my view, this should always be used if a user wants to decode streamed bytes into strings.